contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,404 @@
+2. Set the `VolumeAttachment.Status.AttachError` field detailing the error.
+3. Create an event against the Kubernetes API associated with the `VolumeAttachment` object to inform users what went wrong.
+
+The external-attacher may implement it’s own error recovery strategy, and retry as long as conditions specified for attachment above are valid. It is strongly recommended that the external-attacher implement an exponential backoff strategy for retries.
+
+The detach operation will be triggered by the deletion of the `VolumeAttachment` Kubernetes API objects. The `VolumeAttachment` object will have an `external-attacher` finalizer which will wait for confirmation from the external-attacher before deleting the object.
+
+Once all the following conditions are true, the external-attacher should call `ControllerUnpublishVolume` against the CSI volume driver to detach the volume from the specified node:
+1. A `VolumeAttachment` Kubernetes API object is deleted by Kubernetes: the value for the `VolumeAttachment.metadata.deletionTimestamp` field is set.
+2. The `VolumeAttachment.Status.IsAttached` value in that object is set to `true` (indicating the volume is attached).
+
+If the operation completes successfully, the external-attacher will:
+1. Remove the `external-attacher` finalizer from the list of finalizers on the `VolumeAttachment` object permitting the delete operation to continue.
+
+If the operation fails, the external-attacher will:
+
+1. Ensure the `VolumeAttachment.Status.IsAttached` field remains true to indicate the volume is not yet detached.
+2. Set the `VolumeAttachment.Status.DetachError` field detailing the error.
+3. Create an event against the Kubernetes API associated with the `VolumeAttachment` object to inform users what went wrong.
+
+The new API object called `VolumeAttachment` will be defined as follows:
+
+```GO
+// VolumeAttachment captures the intent to attach or detach the specified
+// volume to/from the specified node.
+//
+// VolumeAttachment objects are non-namespaced.
+type VolumeAttachment struct {
+  metav1.TypeMeta `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+  // Specification of the desired attach/detach volume behavior.
+  Spec VolumeAttachmentSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+  // Most recently observed status of the VolumeAttachment request.
+  // This data may not be up to date.
+  Status VolumeAttachmentStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+
+type VolumeAttachmentSpec struct {
+  // Attacher indicates the name of the volume plugin that MUST handle this
+  // request. This is the name returned by GetPluginName().
+  Attacher string `json:"attacher" protobuf:"bytes,1,opt,name=attacher"`
+
+  // VolumeSource indicates the volume to attach.
+  Volume VolumeSource `json:",inline" protobuf:"bytes,2,opt,name=volumeSource"`
```

Comment:
Good point. But this gets messy. We end up adding a namespace to a non-namespaced object. And the problem with that is multiple pods may reference the same volume, and the secret required for Attach (`ControllerPublishVolume`) should not be tied to the namespace of any one of those pods.

Really what we need is separate secrets for Attach and Mount in `CSIVolumeSource`:

```GO
  // Optional credentials passed to `NodePublishVolume` (secret must exist in the pod namespace).
  MountSecretRef *LocalObjectReference `json:"secretRef,omitempty" protobuf:"bytes,3,opt,name=secretRef"`

  // Optional credentials passed to `ControllerPublishVolume`.
  AttachSecretRef *ObjectReference `json:"secretRef,omitempty" protobuf:"bytes,4,opt,name=secretRef"`
```

This way it is clear that there are a set of credentials to attach the volume (potentially shared between multiple pods), and a set of credentials to mount the volume to the pod (potentially different for each pod).

The problem with this is that a malicious user could point `AttachSecretRef` at any secret in the cluster. But, because CSI volume drivers are responsible for fetching the secret, meaning they have access to all secrets anyway, this may be ok--it gives non-cluster admin no additional privileges.

Thoughts?

Reply:
+@liggitt

Reply:
in the case of a PV, the credential needed to mount a volume to a pod may not live in the same namespace as the pod (generally, I wouldn't expect it to)

Reply:
> CSI volume drivers are responsible for fetching the secret, meaning they have access to all secrets anyway

Not necessarily. If the CSI driver fetches from the API, you can partition credentials by volume type / provisioner and only give the provisioner access to that namespace.

Alternately, if credentials are passed into the CSI API, the node can obtain the secret and provide it as part of the CSI API call.

Reply:
Reading @liggitt's comment https://github.com/kubernetes/community/pull/1258#discussion_r146337244, this will actually work very nicely:

* Provisioner secrets
  * Specified with namespace inside StorageClass
  * external-provisioner access to secrets limited to a specific namespace for that provisioner.
* Attacher secrets
  * Specified with namespace inside CSIVolumeSource (as defined above)
  * external-attacher access to secrets limited to a specific namespace for that attacher.
* Mount secrets
  * Specified without namespace inside CSIVolumeSource (as defined above) --  Mount secrets must live in the user's namespace,
  * Since these will be dereferenced by kubelet, kubelet already has access to pod secrets when preparing (and mounting volumes for) pods.

I will update the spec to reflect this.

Reply:
> We end up adding a namespace to a non-namespaced object. And the problem with that is multiple pods may reference the same volume and the secret required for Attach (ControllerPublishVolume) should not be tied to the namespace of any one of those pods.

If multiple pods use the same PV they are all from the same namespace. I would not  worry about inline volumes much here - let's use credentials from namespace of the first pod that uses the volume.

I see a different problem: we have two sets of secrets.

1. Generic CSI secret(s) so CSI driver can authenticate+authorize **itself** to a storage backend. For example AWS API keys as we use them today - one key for whole AWS cloud provider. We do not need to worry about these secrets, a pod with CSI driver can live in a dedicated namespace and have its own Secrets instance there that gets propagated to the pod as usual (e.g. via env. variables). No problem here.

2. "End user credentials used to authenticate/authorize" various operations (provisioning, attachment, mounting) from CSI spec. The spec does not cover any details about them, "end user credentials" is all we have.
    * These credentials can either authenticate **Kubernetes user** - but do we have a place to store secrets for various CSI drivers for each user? AFAIK we do not tie objects to users. In addition, how can an external provisioner reliably find the right user? All an external provisioner sees is a PVC with some annotations and they can be set/modified by the user.
    * Or they can authenticate **Kubernetes namespaces** - which is IMO the case here. We need either a namespaced secret or some dedicated namespace where all secrets will be stored and we can map them to individual namespaces.
    * Or we won't support end user credentials in Kubernetes. IMO, it's good for alpha - that's how all internal volume plugins work now.

Reply:
> Or we won't support end user credentials in Kubernetes. IMO, it's good for alpha - that's how all internal volume plugins work now.

Except Flex. It can use secret in pod's namespace for mounting :-(.

Reply:
I want to make sure we captured the notes from our October 24 meeting, so here they are:
* For security reasons (to prevent a user from pointing to any secret and getting access to attach and potentially mount a volume they don’t have rights to), only have CSI Volume Source for PVs-- NO inline CSI volume source (can reassess after alpha)
* Rename CSIVolumeSource -> CSIPersistentVolumeSource
* Call out that if a malicious provisioner can obtain an arbitrary secret by setting the mount secret in PV object to whatever secret it wants.
* Have both AttachSecret and MountSecret specify both namespace and secret name.
* Use SecretReference instead of ObjectReference.
* Make sure that as part of CSI Volume Source work that the NodeAuthorizer is updated to allow kubelet read access to Mount Secrets.
* For the future consider for sig-storage
  * Easier UX for pre provision “but only bind to certain namespaces” 

The document has been updated to capture the above.

As to the concern @jsafrane raised, agreed, we can hold off for alpha. Opened https://github.com/container-storage-interface/spec/issues/140 to track this on it on the CSI side.